README.md
[![Actions Status](https://github.com/kfly8/p5-Contextual-Diag/workflows/test/badge.svg)](https://github.com/kfly8/p5-Contextual-Diag/actions) [![Coverage Status](https://img.shields.io/coveralls/kfly8/p5-Contextual-Diag/master.svg?style=flat)](https://coveralls.io/r/kfly8/p5-Contextual-Diag?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/Contextual-Diag.svg)](https://metacpan.org/release/Contextual-Diag)
# NAME

Contextual::Diag - diagnosing perl context

# SYNOPSIS

```perl
use Contextual::Diag;

if (contexual_diag) { }
# => warn "evaluated as BOOL in SCALAR context"

my $h = { key => contexual_diag 'hello' };
# => warn "wanted LIST context"
```

# DESCRIPTION

Contextual::Diag is a tool for diagnosing perl context.
The purpose of this module is to make it easier to learn perl context.

## contexual\_diag()

```perl
contexual_diag(@_) => @_
```

By plugging in the context where you want to know, indicate what the context:

```perl
# CASE: wanted LIST context
my @t = contexual_diag qw/a b/
my @t = ('a','b', contexual_diag())

# CASE: wanted SCALAR context
my $t = contexual_diag "hello"
scalar contexual_diag qw/a b/
```

# LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

kfly8 <kfly@cpan.org>


Changes
Revision history for Perl extension Contextual-Diag

{{$NEXT}}

0.04 2021-05-13T14:09:56Z
    - feat: add shortcut `CDD` #2

0.03 2021-05-10T01:18:16Z
    - eliminate dependence on Contextual::Return #1

0.02 2021-05-08T05:26:04Z
    - fixed typo

0.01 2021-05-08T02:02:36Z

    - original version



lib/Contextual/Diag/Value.pm
package Contextual::Diag::Value;
use 5.010;
use strict;
use warnings;

our $VERSION = "0.02";

use Scalar::Util ();

my %DATA;
my %OVERLOAD;

sub new {
    my ($class, $value, %overload) = @_;

    # Use inside-out to prevent infinite recursion
    my $self = bless \my $scalar => $class;
    my $id = Scalar::Util::refaddr $self;
    $DATA{$id} = {
        value    => $value,
        overload => \%overload,
    };
    return $self;
}

BEGIN {
    my %CONTEXT_MAP = (
        q{""}  => 'STR',
        '0+'   => 'NUM',
        'bool' => 'BOOL',
        '${}'  => 'SCALARREF',
        '@{}'  => 'ARRAYREF',
        '&{}'  => 'CODEREF',
        '%{}'  => 'HASHREF',
        '*{}'  => 'GLOBREF',
    );

    %OVERLOAD = map {
        my $context = $CONTEXT_MAP{$_};

        $_ => sub {
            my $self = shift;

            my $id    = Scalar::Util::refaddr $self;
            my $data  = $DATA{$id};
            my $code  = $data->{overload}->{$context};
            my $value = $data->{value};
            return $code->($value);
        }
    } keys %CONTEXT_MAP,
}

use overload %OVERLOAD, fallback => 1;

sub can {
    my ($invocant) = @_;
    if (ref $invocant) {
        our $AUTOLOAD = 'can';
        goto &AUTOLOAD;
    }
    return $invocant->SUPER::can(@_[1..$#_]);
}

sub isa {
    my ($invocant) = @_;
    if (ref $invocant) {
        our $AUTOLOAD = 'isa';
        goto &AUTOLOAD;
    }
    return $invocant->SUPER::isa(@_[1..$#_]);
}

sub AUTOLOAD {
    my $self = shift;
    our $AUTOLOAD;

    unless (ref $self) {
        die "cannot AUTOLOAD in class call"
    }

    my $obj = do {
        my $id = Scalar::Util::refaddr $self;
        my $data  = $DATA{$id};
        my $code  = $data->{overload}->{OBJREF};
        my $value = $data->{value};
        $code->($value);
    };

    my ($method) = $AUTOLOAD =~ m{ .* :: (.*) }xms ? $1 : $AUTOLOAD;
    return $obj->$method(@_);
}

sub DESTROY {
    my $self = shift;
    my $id   = Scalar::Util::refaddr $self;
    delete $DATA{$id};
    return;
}

1;
__END__

=encoding utf-8

=head1 NAME

Contextual::Diag::Value - wrapping scalar value for diagnostics

=head1 SYNOPSIS

    use Contextual::Diag::Value;

    my $value = 'hello';
    my $v = Contextual::Diag::Value->new($value,
        BOOL => sub { warn 'evaluated as BOOL';  return $_[0] },
    );

    if ($v) { }
    # => warn 'evaluated as BOOL';

=head2 new($value, %overload)

Constructor for Contextual::Diag::Value:

    Contextual::Diag::Value->new($_[0],
        BOOL      => sub { $_[0] },
        NUM       => sub { $_[0] || 0   },
        STR       => sub { $_[0] || ""  },
        SCALARREF => sub { defined $_[0] ? $_[0] : \"" },
        ARRAYREF  => sub { defined $_[0] ? $_[0] : [] },
        HASHREF   => sub { defined $_[0] ? $_[0] : {} },
        CODEREF   => sub { defined $_[0] ? $_[0] : sub { } },
        GLOBREF   => sub { defined $_[0] ? $_[0] : do { no strict qw/refs/; my $package = __PACKAGE__; \*{$package} } },
        OBJREF    => sub { defined $_[0] ? $_[0] : bless {}, __PACKAGE__ },
    );

=head2 OTHER METHODS

=head3 can

Override C<can> to hook OBJREF.

=head3 isa

Override C<isa> to hook OBJREF.

=head1 SEE ALSO

L<Contextual::Return>

=head1 LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

kfly8 E<lt>kfly@cpan.orgE<gt>

=cut



lib/Contextual/Diag.pm
package Contextual::Diag;
use 5.010;
use strict;
use warnings;

our $VERSION = "0.02";

use Exporter;
our @ISA = qw/Exporter/;
our @EXPORT = qw/contextual_diag/;

use Carp ();
use Contextual::Diag::Value;

sub contextual_diag {

    if (wantarray) {
        _diag('wanted LIST context');
        return @_;
    }
    elsif(!defined wantarray) {
        _diag('wanted VOID context');
        return;
    }
    else {
        _diag('wanted SCALAR context');

        return Contextual::Diag::Value->new($_[0],
            BOOL      => sub { _diag('evaluated as BOOL in SCALAR context');  return $_[0] },
            NUM       => sub { _diag('evaluated as NUM in SCALAR context');   return $_[0] || 0   },
            STR       => sub { _diag('evaluated as STR in SCALAR context');   return $_[0] || ""  },
            SCALARREF => sub { _diag('scalar ref is evaluated as SCALARREF'); return defined $_[0] ? $_[0] : \"" },
            ARRAYREF  => sub { _diag('scalar ref is evaluated as ARRAYREF');  return defined $_[0] ? $_[0] : [] },
            HASHREF   => sub { _diag('scalar ref is evaluated as HASHREF');   return defined $_[0] ? $_[0] : {} },
            CODEREF   => sub { _diag('scalar ref is evaluated as CODEREF');   return defined $_[0] ? $_[0] : sub { } },
            GLOBREF   => sub { _diag('scalar ref is evaluated as GLOBREF');   return defined $_[0] ? $_[0] : do { no strict qw/refs/; my $package = __PACKAGE__; \*{$package} } },
            OBJREF    => sub { _diag('scalar ref is evaluated as OBJREF');    return defined $_[0] ? $_[0] : bless {}, __PACKAGE__ },
        );
    }
}

sub _diag {
    local $Carp::CarpLevel = 2;
    goto &Carp::carp;
}

1;
__END__

=encoding utf-8

=head1 NAME

Contextual::Diag - diagnosing perl context

=head1 SYNOPSIS

    use Contextual::Diag;

    if (contextual_diag) { }
    # => warn "evaluated as BOOL in SCALAR context"

    my $h = { key => contextual_diag 'hello' };
    # => warn "wanted LIST context"

=head1 DESCRIPTION

Contextual::Diag is a tool for diagnosing perl context.
The purpose of this module is to make it easier to learn perl context.

=head2 contextual_diag()

    contextual_diag(@_) => @_

By plugging in the context where you want to know, indicate what the context:

    # CASE: wanted LIST context
    my @t = contextual_diag qw/a b/
    my @t = ('a','b', contextual_diag())

    # CASE: wanted SCALAR context
    my $t = contextual_diag "hello"
    scalar contextual_diag qw/a b/

=head1 LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

kfly8 E<lt>kfly@cpan.orgE<gt>

=cut

